### PR TITLE
DR: Updates to stability

### PR DESCRIPTION
**Pull request by gentbot**
Changed `loading` to a more descriptive `isWorking`. This makes is a bit more clear what is happening in the app. Change is primarily for other developers and code readability.

Removed some `useEffect` calls. This was done to avoid running that lifecycle hook when it wasn't needed. The guard clause of `isWorking` was a good protection, but it is good to avoid a call whenever possible.

Changed some string concatenation over to use template literals (widely supported 97%) as they get past the potential edge case of a string with two number next to each other. When using the `+=` operation there is a chance js might perform arithmetic instead of concatenation.

Changed the setting of isWorking to be more deliberate and tied to specific operations. This would make it easier to add other operations is the future that also make use of `isWorking` without the worry of collisions or operations being run when not needed.

Added an error reporter. In the case where some operation does not exist there will be a log of it displayed on the page along with some color change of the operation in question. The error display is limited to one but will show how many times the operation would have been performed. *Note: One error was left in a method name for the scramble option as a demo of this functionality.

Users can remove any operation from the list as long as work is not being done.

Ensured that only positive integers can be entered as a length of the random array.

Made sure all states are initialized with some value to avoid any unexpected bugs or strange ui behavior based on those items at startup.

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -49,30 +51,71 @@ const arrayChangingFunctions = {
...
 
+const operationErrors = new Map();
+
 const App = () => {
-  const [randomArray, setRandomArray] = useState();
+  const [randomArray, setRandomArray] = useState([]);
   const [outputArray, setOutputArray] = useState([]);
   const [changes, setChanges] = useState([]);
-  const [loading, setLoading] = useState();
+  const [isWorking, setIsWorking] = useState(false);
```
Comment (pyreta) [inline src/App.js:60]: why did you choose `isWorking` here?  Does this mean the app "is working" as in "it is working properly" or as in "it is working on completing an operation?"

File: src/App.js [hunk 2]
```diff
@@ -49,30 +51,71 @@ const arrayChangingFunctions = {
...
+  const applyCurrentChanges = () => {
+    setOutputArray(changes.reduce((output, change) => {
+      const { method, fn } = arrayChangingFunctions[change];
+      if(typeof output[method] !== 'function') {
+        const existingOperation = operationErrors.has(change);
+        const errObj = {
+          count: existingOperation ? (operationErrors.get(change).count + 1): 1,
+          name: method,
+          changeType: change,
+          message: `Arrays do not contain the method \"${method}\" this operation was skipped`
```
Comment (pyreta) [inline src/App.js:72]: do double quotes need to be escaped in an interpolated string?

File: src/App.js [hunk 3]
```diff
@@ -49,30 +51,71 @@ const arrayChangingFunctions = {
   },
 };
 
+const operationErrors = new Map();
```
Comment (pyreta) [inline src/App.js:54]: cool!

**Review by pyreta — COMMENTED** [review]
Nice work!
You addressed a lot of things and made some interesting and worthwhile changes.
I left a couple of comments but my main question is about optimization.
Which part of this app could most benefit from optimization and how would you do it?  (You don't need to write more code if you don't want to.  A simple plain english explanation will suffice).
Again, great work!

File: src/App.js [hunk 4]
```diff
@@ -6,19 +6,21 @@ export const makeRandomString = (length) => {
...
-      newLetters += letters[i].toLowerCase();
-    }
+  /*
+  * safer to use string interpolation over the += operator.
+  * If there are two numbers next to each other that operation might 
+  * do a math calculation instead of concatenating the strings
+  */
+  for(const letter of letters) {
+    const newCase = letter === letter.toUpperCase()? letter.toLowerCase() : letter.toUpperCase()
+    newLetters = `${newLetters}${newCase}`;
```
Comment (salanki) [inline src/App.js:23]: There is a performance impact here though. Collecting them into an array and then joining the array at the end is going to involve less string generation.

**Pull request closed without merging**